File: src/python/xerxes/memory/plugins/supermemory.py
# Copyright 2026 The Xerxes-Agents Author @erfanzar (Erfan Zare Chavoshi).
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""Supermemory cloud memory provider (https://supermemory.ai)."""

from __future__ import annotations

import os
from typing import Any

import httpx

from ._base import ExternalMemoryProviderBase


class SupermemoryProvider(ExternalMemoryProviderBase):
    """Memory provider backed by supermemory.ai's HTTP API.

    Requires ``SUPERMEMORY_API_KEY``."""

    name = "supermemory"
    namespace_label = "super"
    required_env = ("SUPERMEMORY_API_KEY",)

    BASE_URL = "https://api.supermemory.ai/v1"

    def _client(self) -> httpx.Client:
        """Return an authenticated ``httpx.Client`` for the Supermemory API."""
        return httpx.Client(
            base_url=self.BASE_URL,
            headers={"Authorization": f"Bearer {os.environ['SUPERMEMORY_API_KEY']}"},
            timeout=30.0,
        )

    def _call_upstream(self, action: str, arguments: dict[str, Any]) -> Any:
        """Translate a standard action to the matching Supermemory endpoint."""
        with self._client() as c:
            if action == "add":
                resp = c.post("/memories", json={"content": arguments["content"]})
            elif action == "search":
                resp = c.post("/memories/search", json={"q": arguments["query"]})
            elif action == "list":
                resp = c.get("/memories", params={"limit": arguments.get("limit", 20)})
            elif action == "remove":
                resp = c.delete(f"/memories/{arguments['entry_id']}")
            else:
                raise ValueError(f"unknown supermemory action: {action}")
            resp.raise_for_status()
            return resp.json() if resp.content else {"ok": True}


PROVIDER = SupermemoryProvider()

__all__ = ["PROVIDER", "SupermemoryProvider"]

File: src/typescript/src/memory/plugins/supermemory.ts
// Copyright 2026 The Xerxes-Agents Author @erfanzar (Erfan Zare Chavoshi).
// Licensed under the Apache License, Version 2.0.

import {
  ExternalMemoryProviderBase,
  memoryPluginLimit,
  requiredMemoryPluginArgument,
  requiredMemoryPluginEnvironment,
  type MemoryPluginEnvironment,
} from './base.js'
import { HttpMemoryUpstream, memoryPluginPathSegment, memoryPluginUrl, type MemoryPluginHttpTransport } from './http.js'

/** Default Supermemory cloud API endpoint. */
export const DEFAULT_SUPERMEMORY_BASE_URL = 'https://api.supermemory.ai/v1'

/** Explicit Supermemory configuration and host-owned HTTP transport. */
export interface SupermemoryProviderOptions {
  readonly baseUrl?: string
  readonly clock?: () => number
  readonly environment: MemoryPluginEnvironment
  readonly transport: MemoryPluginHttpTransport
}

/** Supermemory HTTP adapter with no implicit client or network fallback. */
export class SupermemoryProvider extends ExternalMemoryProviderBase {
  readonly baseUrl: string

  constructor(options: SupermemoryProviderOptions) {
    const baseUrl = options.baseUrl ?? DEFAULT_SUPERMEMORY_BASE_URL
    super({
      name: 'supermemory',
      namespaceLabel: 'super',
      requiredEnvironment: ['SUPERMEMORY_API_KEY'],
      environment: options.environment,
      upstream: new HttpMemoryUpstream({
        providerName: 'supermemory',
        transport: options.transport,
        requestFor: (action, arguments_) => {
          const apiKey = requiredMemoryPluginEnvironment(options.environment, 'SUPERMEMORY_API_KEY')
          const headers = { Authorization: `Bearer ${apiKey}` }
          if (action === 'add') {
            return {
              method: 'POST',
              url: memoryPluginUrl(baseUrl, 'memories'),
              headers,
              body: { content: requiredMemoryPluginArgument(arguments_, 'content') },
            }
          }
          if (action === 'search') {
            return {
              method: 'POST',
              url: memoryPluginUrl(baseUrl, 'memories/search'),
              headers,
              body: { q: requiredMemoryPluginArgument(arguments_, 'query') },
            }
          }
          if (action === 'list') {
            return {
              method: 'GET',
              url: memoryPluginUrl(baseUrl, 'memories', { limit: memoryPluginLimit(arguments_, 20) }),
              headers,
            }
          }
          const entryId = requiredMemoryPluginArgument(arguments_, 'entry_id')
          return {
            method: 'DELETE',
            url: memoryPluginUrl(baseUrl, `memories/${memoryPluginPathSegment(entryId)}`),
            headers,
          }
        },
      }),
      ...(options.clock === undefined ? {} : { clock: options.clock }),
    })
    this.baseUrl = baseUrl
  }
}
